### ISSUE TITLE
Proposal for controller security (policies)

### ISSUE THREAD

**Issue by tapuzzo-fsi** [marklogic/roxy#615]
Hi,

I don't like having to edit my controllers to insert security logic in every single controller function. I have a simple controller security proposal.  The idea is to introduce a concept called a policy.  The policy is an XQuery statement (not a module) that is executed prior to invoking the controller function.  If the policy (or policies) throw an exception, then the request is denied (with an appropriate HTTP response.) If all the configured policies complete sucessfully (regardless of return value), then the controller is invoked normally.

I have a sample implementation that I could push to a feature branch on my fork or make a pull-request if anyone is interested.  Here is an example of what I'm proposing:

src/app/config/config.xqy:
```xquery
  ...
  <routes>
    <request resource="status" >
      <rest:policy name="isLoggedIn"/>
      <!-- Could list as many policies as you want here -->
    </request>
    <request resource="asset" >
      <rest:policy name="isLoggedIn"/>
      <rest:policy method="GET" name="canRead"/>
      <rest:policy method="PUT" name="canWrite"/>
      <rest:policy method="POST" name="canWrite"/>
      <rest:policy method="DELETE" name="canWrite"/>
    </request>
  </routes>
  ...
```

src/app/policies/isLoggedIn.xqy:
```xquery
if (fn:exists(xdmp:get-request-username())
then ()
else fn:error(xs:QName("UNAUTHORIZED"))
```

src/app/policies/canRead.xqy:
```xquery
if (example:current-user-can-read())
then ()
else fn:error(xs:QName("FORBIDDEN"))
```

src/app/policies/canWrite.xqy:
```xquery
if (example:current-user-can-write())
then ()
else fn:error(xs:QName("FORBIDDEN"))
```

This approach was inspired by the [Sails](http://sailsjs.org/) nodejs web framework's [Policies](http://sailsjs.org/documentation/concepts/policies) design.

This model could be extended to be more like an ExpressJS middleware where the response could be augmented by a policy, but that would be a lot more complex I think.  I'm open to discussion about ways to make this more robust or simpler.  Please let me know what you think.

**Comment by dmcassel** [comment]
My first thought is: why aren't MarkLogic roles sufficient for this? With properly defined roles applied to users, a user trying to read content for which he/she doesn't have read permission will not get anything back. A user trying to update content for which he/she doesn't update permission will get rejected. I'm concerned this approach might be used in place of proper database-level security. 

Moving past that question, it sounds like this would require some work in the src/roxy directory, in order to apply the policies. The code that applies the policies must gracefully handle finding no policies (not only because someone might choose not to supply any, but also because someone who upgrades their Roxy files (`ml upgrade -h`) won't have any policies defined

**Comment by grtjn** [comment]
I agree with Dave that it makes less sense to use policies that way to check access and such, but it would be a great hook to tap in to execute privileges. I could imagine having an edit page. Instead of having doc permissions preventing a user to save edits, it would be good to stop the user upfront by telling he/she can't access the edit page..

**Comment by tapuzzo-fsi** [comment]
When using application-level security, I need to give read access to static resources, and some simple queries, but don't want to give access to (all) the content.  I have found it difficult to trap the execute permissions or read-only permissions without causing 500 errors when I really want 401/403 errors.  With this approach there is a well-defined way to convert security errors into HTTP error codes (my initial implementation is very simplistic, but would be straightforward to improve.) 

Using roles and permissions in the database requires that files be deployed into the database with differing permissions and this is very difficult to manage and debug.  In addition, if you hard-code rules into each controller function then it is much more challenging to manage change over time.  

Another advantage of this policy approach allows the rules to be dynamic without making controllers more complex.  It is independent of unit testing but integration tests still enforce policies.

Security is generally considered a cross-cutting concern and best practice is to store security logic separate from other business logic.

I have another idea triggered by @grtjn specifically for an even simpler approach to fine-grain execute permissions, I will write that up as soon as I can.  It is an alternative to what I described above but I need to validate my idea before I write it up.

**Comment by dmcassel** [comment]
I would expect that you'd give read and execute permission liberally in the modules database (which includes static resources like html/css/js/images), then be more particular about read/insert/update permissions in the content database. 

I agree that `ml {env} deploy content` doesn't give much control over different permissions for different content, but between Query Console and CORB I'd think it's pretty straightforward to update those permissions as needed. While this is a little work, it's pretty important. Handling security in the application code makes it much more likely that a mistake allows incorrect access or update; if it's handled in the database layer, access through *any* channel is properly controlled. 

I guess I'm having a hard time understanding a scenario where application code is a better place to manage security than the use of roles and permissions.

**Comment by tapuzzo-fsi** [comment]
I am not sure I understand how to implement what you're describing in a small amount of code.

The project I'm working on has a group of read-only users and a group of editors who can make changes. The default Roxy permissions and roles don't work with read-only users so I had to update the permissions already using `app_specific.rb`. Once I have the permissions and roles set up appropriately, the 'read-only' users cannot make changes no matter what. This is good because the database itself is a backstop to prevent unauthorized modifications. However, the error conditions that arise from lacking permission on a fine-grain level require coding to trap errors in every controller.

The issue of execute permissions can be handled by using `xdmp:security-assert`, but again if there is no support from Roxy for simplifying this, then you need 5-10 lines of code to wrap every controller function and there is more opportunity for inconsistency in the error response and enforcement.

Using my proposed framework, the execute privileges can be enforced declaratively by sending the policy as an external variable... I'm also guessing that I shouldn't be using the `rest` namespace in my extensions, but that's a minor niggle:

```xquery
...
  <routes>
    <request resource="status" >
      <rest:policy name="hasExecutePrivilege">
        <rest:privilege>http://foo.bar.baz</rest:privilege>
      </rest:policy>
    </request>
  </routes>
  ...
```
app/policies/hasExecutePrivilege.xqy
```xquery
xquery version "1.0-ml";
declare namespace rest = "http://marklogic.com/appservices/rest";
declare variable $rest:policy as element(rest:policy) external;
try {
  xdmp:security-assert($rest:policy/rest:privilege, "execute")
} catch ($e) {
  fn:error(xs:QName('FORBIDDEN'), 'FORBIDDEN', $e/*:format-string)
}
```

Maybe I'm not being ambitious enough :-)

Instead of thinking of this as a purely security related concept, the policies approach could be extended to generic cross-cutting pre/post wrappers. With the middleware concept, I would think that these middleware functions would have access to the controller-helper value map so that values for the views and layouts could be injected and changed before and/or after the controller runs.

Thanks for considering this, I can tell people are skeptical.

**Comment by joemfb** [comment]
It's worth noting that the XQuerrail Framework includes interceptors, which, I believe, are a generic middleware implementation: https://github.com/nativelogix/xquerrail2.framework/wiki/Interceptors

Their getting-started page mentions an ml-security interceptor as well: https://github.com/nativelogix/xquerrail2.framework/wiki/Getting-Started

**Comment by tapuzzo-fsi** [comment]
If anyone wants to try this out, I've committed an initial implementation into my fork under the feature branch [feature/mvc-policies](https://github.com/tapuzzo-fsi/roxy/tree/feature/mvc-policies).

**Comment by dmcassel** [comment]
I'll have to come back to this, but just to be clear -- I'm not rejecting what you say, just trying to make sure it's not a substitute for in-database security. 

Would it work to use amps in cases where you want certain roles to have limited update capabilities? (Is that the use case?)

**Comment by tapuzzo-fsi** [comment]
(Apologies for the close/open, I am new to github and clicked the wrong button.)

Originally I wrote this because application-level security requires the 'default-user' to be a higher-powered user than I would like in order for `ml deploy modules` to work.  So I had set it up so that the default-user had more mojo and I needed a simple way to nerf that user so they couldn't do updates anonymously.  

I learned about using `app_specific.rb` to override the default permissions assigned to modules during deploy and then figured out swapping out `admin` with `${default-user}` during the `rest-deploy` (I don't like that this is a race-condition security hole, but adequate for my needs.)

With those changes, it is easier to lock-down things using baseline MarkLogic security, which is good.

My remaining use-case is the situation where I want to restrict access to a controller that users would otherwise have access to. Things like running a slow report or restricting an endpoint that receives ad-hoc XQuery that it will xdmp:eval().  Of course this can be done in the controller code itself, but I like the idea of having a cross-cutting feature to enforce security rules outside the code for implementation.

I am also using it to provide a short-circuit to catch errors early so that the resulting error messages are easier to get correct for the client.  When a security error is encountered deep in the code, it ends up with a 500 to the client instead of 401/403.... of course a global error catch could also be used for this... that's another topic but I should look into it.
